### ISSUE TITLE
BUG: freezeAllScrolls does not work with native Android scrolling

### ISSUE THREAD

**Issue by rolandoldengarm** [driftyco/ionic#5205]
*Android 4.4*

With native scrolling enabled (which is default), the following calls have no effect:
````
$ionicScrollDelegate.getScrollView().options.scrollingY = false;
$ionicScrollDelegate.freezeAllScrolls(true);
$ionicScrollDelegate.freezeScroll(true);
````
My HTML consists of <ion-content> with an <ion-list> and an <ion-refresher>

With JS scrolling enabled, it works.

**Comment by embeddedwes** [comment]
I have confirmed this issue on Android 4.4.4. Looking now for a potential solution.

**Comment by rolandoldengarm** [comment]
There are more scrolling issues with native Android. E.g. scrollComplete never fires.

**Comment by talamaska** [comment]
@rolandoldengarm, scrollComplete doesn't fire in chrome emulator as well.

Freeze scrolling doesn't work too.

**Comment by rolandoldengarm** [comment]
Any potential solution @embeddedwes ?

**Comment by bodinaren** [comment]
**Alternative 1:**
I found that you could freeze the scroll with a somewhat hidden function called `freezeScrollShut(shouldFreezeShut)`. Unfortunately it's only available on the objects in `$ionicScrollDelegate._instances`. So here's what I did:
```js
$ionicScrollDelegate._instances.forEach(function(scrollDelegate) {
    if (scrollDelegate.$$delegateHandle === "myHandle") { // I set delegate-handle on the ion-content in question.
        scrollDelegate.freezeScrollShut(true);
        scrollDelegate.freezeScroll(true); // You'll still want this for where jsScrolling is in effect, e.g. iOS and Windows Phone.
    }
});
```
Note that `_instances` is considered a private array, so this could break with new versions without any notice.

**Alternative 2:**
Before finding this, I wrote a directive that also locks scrolling. The only downside at this point is that there's a flash when you re-add `overflow: visible` as it scrolls to the top and then scrolls back down to the original position. I also haven't done any extensive testing yet, as I decided to go with alternative 1 in the end. Anyway, if anyone wanna try it and possibly improve it, here it is:
```js
.directive("freezeScroll", ["$ionicScrollDelegate", function ($ionicScrollDelegate) {
    return {
        restrict: "A",
        link: function (scope, element, attrs) {
            
            if (!attrs.delegateHandle) throw "htFreezeScroll directive requires attribute delegate-handle to be set";

            var scroll = element.children(".scroll")[0],
                scrollDelegate = $ionicScrollDelegate.$getByHandle(attrs.delegateHandle);

            scope.$watch(function () { return scope.$eval(attrs.htFreezeScroll) }, function (newVal, oldVal) {
                if (newVal === oldVal) return;

                var scrollTop;

                if (newVal) {
                    scrollTop = scrollDelegate.getScrollPosition().top;

                    angular.element(scroll).css("overflow", "hidden");
                    scroll.scrollTop = scrollTop;
                } else {
                    scrollTop = scroll.scrollTop;

                    angular.element(scroll).css("overflow", "visible");
                    scrollDelegate.scrollTo(0, scrollTop, false);
                }
            });
        }
    };
}]);
```
```html
<ion-content freeze-scroll="{{scopeVariable}}" delegate-handle="myScroll">
```

**Comment by talamaska** [comment]
@bodinaren great suggestions, but we are not supposed to use some private internal apis, and we don't have to write directives in order to fix an existing feature which doesn't work.

**Comment by bodinaren** [comment]
@talamaska Of course. I'm not saying this is how it should be, I just offered some alternatives for people who don't want to, or can't, wait for it to be fixed (it was at least over a month since this issue was raised after all). And maybe it'll also help someone in finding a proper solution for it.

**Comment by jgw96** [comment]
Hey everyone! So, im having trouble reproducing this issue and would love if someone could create a codepen that shows this issue! I am testing with my nexus 6 running the android n beta preview with ionic 1.2.4 and the above `$ionicScrollDelegate` methods seem to be working for me. Thanks!

**Comment by talamaska** [comment]
https://codepen.io/zlatip/pen/qZprxP

**Comment by jgw96** [comment]
Hello everyone! This has been fixed and will be in the next release, Ionic 1.3! Also, if you would like to do some testing with 1.3 then you can follow these instructions here: https://forum.ionicframework.com/t/ionic-1-3-beta/49133 .

**Comment by talamaska** [comment]
I tested again, the content still scrolls without any problems. I made sure that js scrolling is disabled and i have set properly the scroll handle. I replaced the sources. How guys are you testing this? Could you provide an example where it works?

**Comment by danbucholtz** [comment]
@talamaska, sure let me get up to speed on this issue again and I'll upload my example for you to verify.

**Comment by talamaska** [comment]
Tested on a new sample sidemenu app, added more items to playlists, still scrolling without any interruption. Put the freezing in the controller and disabled jsscrolling in config phase

**Comment by danbucholtz** [comment]
Hi @talamaska, I have verified this functions correctly on my end.  This is a bit complicated of a process for you to recreate, and I apologize for that.

Can you execute this steps:

1.  Execute `ionic start verify-native-srolling http://codepen.io/danbucholtz/pen/GZxmqY`
2.  In the verify-native-scrolling www/lib/ionic/js directory, can you replace with `ionic.bundle.js` with this nightly version from here `http://code.ionicframework.com/nightly/js/ionic.bundle.js`
3.  Execute `ionic run android`

The code is very simple.  I just modified one of our starter projects.  Look in app.js for the PlaylistCtrl.  I am just showing a list of items instead, with the scrolling frozen.

```
.controller('PlaylistCtrl', function($scope, $ionicScrollDelegate) {

    $scope.$on("$ionicView.enter", function(event, data){
      var items = [];
      for ( var i = 0; i < 50; i++ ){
        items.push({name: (i +1)});
      }
      $scope.items = items;
      $ionicScrollDelegate.freezeAllScrolls(true);
    });
    
  });
```

Works for me with both JS and Native scrolling on iOS (9.3 on iPhone 6S) and Android (4.4 on Galaxy S4, Moto G).  You can toggle between native and JS scrolling by flipping the `overflow-scroll="true"` boolean on `templates.playlist.html` embedded template in `index.html`

```
<script id="templates/playlist.html" type="text/ng-template">
    <ion-view view-title="Playlist">
      <ion-content has-header="true" overflow-scroll="true">
        <ion-list>
          <ion-item ng-repeat="item in items">
            <h2>Item #{{item.name}}</h2>
          </ion-item>
        </ion-list>
      </ion-content>
    </ion-view>
  </script>
```

I am going to close this.  Please let me know if you continue to have issues and we'll re-open.

Thanks,
Dan

**Comment by talamaska** [comment]
I tested on Samsung Note 4, with Android 5.1.1, without crosswalk your code from code pen the ionic bundle version is v1.2.4-nightly-2959, the Angular inside is 1.5.3 and it is still scrolling.

**Comment by danbucholtz** [comment]
@jgw96, can you test this out please using the instructions I detailed above?

**Comment by jgw96** [comment]
@danbucholtz testing this issue now on my devices.

**Comment by talamaska** [comment]
@jgw96 and?

**Comment by danbucholtz** [comment]
@talamaska, we are seeing this issue still.  You are correct.  The scroll is initially frozen but it gets unfrozen on subsequent "swipes".  We are working on a fix for 1.3.1.  This won't make the 1.3 cut this week.

Thanks,
Dan
